### DISCUSSION TITLE
Enhancement: Create config mnemonics for Image Input and Associated Features

### DISCUSSION THREAD

**Discussion by Moorer624** [lllyasviel/Fooocus#3370] [Q&A] [answered]
I'd like to be able to create presets that let me control
1. Input Image Checkbox Default as Checked.
2. Upscale or Variation Default to Vary Subtle.
3. Image Prompt Tab > Defaults for all 4 image entry points (I almost always enter a single reference image 3 times - as ImagePrompt, as PyraCanny, as FaceSwap to get the best results when using a reference image.

**Comment by mashb1t** [comment]
Feel free to adjust the webui.py as these seem like very individual requirements of yours.

**Reply by Moorer624** [reply]
I would think others would want some control over the default behavior or the advanced features found in the Image Prompt tab and the desire to have the Image Checkbox default as checked. Why do you think this is "very individual requirements"? There are mnemonics for most of the features - like the ones below. Why not make them for other fields? I don't know how to edit webui.py, unfortunately.

"default_refiner_switch": 0.5,
    "default_cfg_scale": 3.0,
    "default_sample_sharpness": 2.0,
    "default_sampler": "dpmpp_sde",
    "default_scheduler": "karras",
    "default_performance": "Speed",
    "default_overwrite_step": 14,
    "default_prompt": "",
    "default_prompt_negative": "",
    "default_styles": [],
    "default_aspect_ratio": "1024*1024",

**Reply by mashb1t** [reply]
Adding images as default is not possible on client side as the Fooocus server needs to have the default images. This only can work if you access Fooocus from the same machine it runs on => individual solution or breach of data privacy / need to know principle.
You can adjust the webui.py by searching for the labels in the website you want to adjust, then set the default value to whatever you desire to match your requirements.
Hope this helped in explaining why at least setting image prompt defaults is not a solution for everybody.

**Reply by Moorer624** [reply]
Oh! I think I might not be communicating this correctly. Here's what I was hoping...I would create a preset that includes the following: 

 "default_refiner_switch": 0.5,
    "default_cfg_scale": 3.0,
    "default_sample_sharpness": 2.0,
    "default_sampler": "dpmpp_sde",
    "default_scheduler": "karras",
    "default_performance": "Speed",
    "default_overwrite_step": 14,
    "default_prompt": "",
    "default_prompt_negative": "",
    "default_styles": [],
    "default_aspect_ratio": "1024*1024",

New Ones:
"default_overwrite_refiner_step" (I'd match the value in this parameter to value in "default_overwrite step")
"default_input_image_checkbox" True or False
"default_image_entry 1" (with a selection of ImagePrompt, PyraCanny, CPDS, FaceSwap) so that each of the 4 areas you can enter an image on the Image Prompt Tab, users could set a default value of one of the existing radio buttons.
"default_image_entry 2"
"default_image_entry 3"
"default_image_entry 4"

Does this make more sense? These parameters would exist as parameters that users could use (if desired) in a preset.

**Reply by mashb1t** [reply]
Thanks for rephrasing, i get what you mean. 
1. default_overwrite_refiner_step doesn't make any sense, as the default_overwrite_step is the maximum and default_overwrite_switch the step count where to switch
2. default_input_image_checkbox is easily doable
3. default_image_entry_1 to 4, as you describe, would be a combined attribute of image path and default_ip => have to be 2 attributes to separate them logically. Issues mentioned above apply.

The thing is: up until now preset keys do not control image prompt fields or inpainting etc. (any image input parameters) as they would also have to be set in metadata / loaded from metadata as presets and metadata use the same logic to apply updates to the UI, which is very inconvenient and not wanted here.
=> This is intended because you can't reconstruct the result by applying these settings only, you still need an additional mask/input image, which is obviously not included in the metadata of the image itself.

In short:
1. No
2. Yes
3. No (with current approach)

Hope this was clearly described and you got the point.

**Reply by Moorer624** [reply]
I hear what you're saying for the most part, and don't want to take up more of your time. One last thought...every time I start Fooocus, I always (a) check the Input Image checkbox, (b) go to the Image Prompt tab and set the radio buttons of the first 3 sections to ImagePrompt, PyraCanny, FaceSwap - whether or not I end up putting any images into those boxes. If not images are entered, then my understanding is that it does nothing, BUT if I do put images in, I almost always put the same image in 3 times, once for each of these 3 radio buttons. Maybe that's the unique part of my workflow? I just get much better images when doing it this way.

**Reply by mashb1t** [reply]
@Moorer624 i've had a second look and it wasn't such a hassle to separate configs from metadata. Implemented in PR https://github.com/lllyasviel/Fooocus/pull/3382

**Reply by mashb1t** [reply, marked as answer]
@Moorer624 merged to main, please restart Fooocus and find the config options in the config_modification_tutorial.txt

**Comment by Moorer624** [comment]
This is absolutely fantastic. I just set it up and it’s working well. I’m not a developer, but I spent 25 years working at a company called Epic (Verona, WI) that does electronic medical records software (#1 in the world) as a Project Director. I incredibly appreciate you looking into this as you did.

I’m a bit new to GitHub and Stable Diffusion in general. I’d like to send you a “tip”. What’s the best way to do that?

Thanks again!

[signature_320236866]

From: Manuel Schmid ***@***.***>
Date: Saturday, July 27, 2024 at 3:49 PM
To: lllyasviel/Fooocus ***@***.***>
Cc: Moorer624 ***@***.***>, Mention ***@***.***>
Subject: Re: [lllyasviel/Fooocus] Enhancement: Create config mnemonics for Image Input and Associated Features (Discussion #3370)

@Moorer624<https://github.com/Moorer624> i've had a second look and it wasn't such a hassle to separate configs from metadata. Implemented in PR #3382<https://github.com/lllyasviel/Fooocus/pull/3382>

**Reply by mashb1t** [reply]
@Moorer624 don't feel obligated to do so, but if you really want to you can find my PayPal [here](https://paypal.me/mashb1t) or a sponsor link [here](https://github.com/sponsors/mashb1t).
